Fix inconsistent subscription status after cancellation with centralized cancellation logic

lmcrean opened the pull request:
## Issue #78 
After cancelling a subscription, the status is inconsistent across different areas (#78):
- Backend Orders: shows "pending subscription" ❌
- My Account > Subscriptions: shows "pending subscription" ❌  
- My Account > Orders: shows "cancelled" ✅

## Root Cause
Subscription cancellation events weren't properly synchronized between orders, subscriptions, and renewal orders. The gateway was missing status propagation logic.

## Solution
Add centralized cancellation handling and status synchronization:

1. **Unified cancellation handler** - Single method to handle all cancellation scenarios (payment cancelled, billing request cancelled, subscription cancelled)

2. **Status synchronization hooks** - Ensure parent orders reflect subscription status changes and vice versa

3. **Metadata cleanup** - Clear pending payment flags when cancellation occurs

## Changes
- `class-wc-gocardless-gateway.php`: Add `handle_subscription_cancellation()` method
- `class-wc-gocardless-gateway-addons.php`: Add `sync_parent_order_status()` hook
- Update webhook handlers to use centralized cancellation logic

## Testing
1. Create subscription order
2. Cancel subscription from frontend/backend
3. Verify all areas show "cancelled" status consistently

lmcrean commented:
Confirmed in unit testing, does anyone have access to trying on local machine with the paid subscription plugin? @diegocurbelo  @peterwilsoncc  @joemcgill  @dkotter

dkotter commented on includes/class-wc-gocardless-gateway-addons.php:
```diff
@@ -33,13 +33,52 @@ public function __construct() {
+	/**
+	 * Synchronize parent order status when all subscriptions are cancelled.
+	 *
+	 * @since 2.9.8
```
```suggestion
	 * @since x.x.x
```

dkotter commented on includes/class-wc-gocardless-gateway.php:
```diff
@@ -2315,11 +2313,58 @@ protected function _process_billing_request_event_cancelled( array $event ) {
+	 * Centralized handler for subscription cancellation.
+	 * Ensures consistent status updates across orders and subscriptions.
+	 *
+	 * @since 2.9.8
```
```suggestion
	 * @since x.x.x
```

iamdharmesh commented on includes/class-wc-gocardless-gateway.php:
```diff
@@ -2066,11 +2068,7 @@ protected function _process_subscription_event( array $event ) {
-				// Only cancel WCS subscriptions when the order missing mandate.
-				$mandate_in_order = $this->get_order_resource( $order_id, 'mandate', 'id' );
-				if ( ! $mandate_in_order && class_exists( 'WC_Subscriptions_Manager' ) ) {
-					WC_Subscriptions_Manager::cancel_subscriptions_for_order( $order_id );
-				}
+				$this->handle_subscription_cancellation( wc_get_order( $order_id ), __( 'Subscription cancelled.', 'woocommerce-gateway-gocardless' ) );
```
We are no longer using GoCardless-side subscriptions, and this code is only for backward compatibility. So, no changes are needed here.

lmcrean replied:
Resolved, thank you

iamdharmesh reviewed (changes requested):
Thanks a lot for the PR, @lmcrean. While reviewing the PR, I noticed that the issue report(#78) didn’t include the background details and all the necessary information (apologies for that). I’ve added additional context and details about the issue [here](https://github.com/gocardless/woocommerce-gateway-gocardless/issues/78#issuecomment-3359940627).

While testing this PR, I also noticed that it doesn’t fully resolve the issue. Could you please review the additional information I’ve shared and confirm?

Thanks again for your contribution!

lmcrean commented:
Thanks @iamdharmesh for the context! I can see that the initial PR handled webhook-based cancellations but missed user-initiated cancellations with unconfirmed payments.

So I've added a fix for the missing piece. This completes the behavior described in your summary:
- Payment confirmed → User cancels → Pending Cancellation
- Payment unconfirmed → User cancels → Cancel immediately (now fixed)

Let me know if you'd like me to adjust anything

iamdharmesh commented on includes/class-wc-gocardless-gateway.php:
```diff
@@ -1921,6 +1921,8 @@ protected function _process_payment_event( array $event ) {
 			case 'cancelled':
 				$new_status = 'cancelled';
+				// Centralized cancellation handling
+				$this->handle_subscription_cancellation( $order, __( 'Payment cancelled.', 'woocommerce-gateway-gocardless' ) );
```
As we have already handled subscription cancellation for webhook in #50. So, this handling is no longer needed unless anything specific not working here.

lmcrean replied:
Thanks for the catch. Removed.

iamdharmesh commented on includes/class-wc-gocardless-gateway-addons.php:
```diff
@@ -33,13 +33,100 @@ public function __construct() {
+			add_action( 'woocommerce_subscription_status_updated', array( $this, 'sync_parent_order_status' ), 10, 3 );
+
+			// Skip pending-cancel for subscriptions with unconfirmed payments
+			add_filter( 'woocommerce_subscription_use_pending_cancel', array( $this, 'maybe_skip_pending_cancel_status' ), 10, 2 );
```
This filter has only 1 argument and due to that I am getting fatal error. Could you please help to check once?

lmcrean replied:
Thanks @iamdharmesh for the review.

I've fixed this by:

1. **Removing the problematic filter hook** (includes/class-wc-gocardless-gateway-addons.php #L41)
2. **Removing the `maybe_skip_pending_cancel_status()` method**
3. **Extended `sync_parent_order_status()` instead** - it already receives the subscription object via the `woocommerce_subscription_status_updated` action

Now when a subscription transitions to `pending-cancel`, the method intercepts it, checks the payment status, and immediately cancels if unconfirmed. Same functionality, no fatal error expected.

lmcrean commented:
thanks @iamdharmesh could i get your review please? Addressed your suggestions

iamdharmesh commented on includes/class-wc-gocardless-gateway-addons.php:
```diff
@@ -33,13 +33,83 @@ public function __construct() {
+
+		// Handle pending-cancel -> cancelled transition for unconfirmed payments
+		if ( 'pending-cancel' === $new_status && 'pending-cancel' !== $old_status ) {
+			$parent_order = $subscription->get_parent();
```
Maybe we should consider checking payment status of last order? 🤔 Thinking cases where subscription is getting cancelled after 2-3 renewals.

lmcrean replied:
Good catch! Changed from `get_parent()` to `get_last_order('all')` with fallback.

iamdharmesh commented on includes/class-wc-gocardless-gateway-addons.php:
```diff
@@ -33,13 +33,83 @@ public function __construct() {
+			$parent_order = $subscription->get_parent();
+			if ( $parent_order ) {
+				// Check payment confirmation status
+				$payment_status = $parent_order->get_meta( '_gocardless_payment_status', true );
```
Maybe we should get payment status from GoCardless directly, to make sure it's latest (to avoid any edge case). what do you think?

lmcrean replied:
I think it's worth doing. 

Might add ~200ms etc. API call during cancellation, but ensures we always have the latest status (avoids webhook delay edge cases). Since this is a user-initiated action and the check is critical (determines if user can undo cancellation), the accuracy is worth the small performance cost.

Implementing API call to get fresh payment status.

iamdharmesh commented on includes/class-wc-gocardless-gateway-addons.php:
```diff
@@ -33,13 +33,83 @@ public function __construct() {
+			'cancelled',
+			__( 'Order cancelled - all subscriptions have been cancelled.', 'woocommerce-gateway-gocardless' )
+		);
+		$parent_order->delete_meta_data( '_gocardless_payment_pending' );
```
I think we are not storing `_gocardless_payment_pending` order meta anywhere, is there any reason to delete this?

lmcrean replied:
Thanks for spotting. Searched the codebase and this metadata is never set anywhere, only deleted. Removed the deletion line.

iamdharmesh commented on includes/class-wc-gocardless-gateway.php:
```diff
@@ -2315,11 +2315,58 @@ protected function _process_billing_request_event_cancelled( array $event ) {
 		if ( empty( $payment ) && empty( $mandate ) ) {
 			$order->update_status( 'cancelled', __( 'Billing request cancelled.', 'woocommerce-gateway-gocardless' ) );
 			$this->update_order_resource( $order, 'billing_request', $billing_request );
+			$this->handle_subscription_cancellation( $order, __( 'Billing request cancelled by customer.', 'woocommerce-gateway-gocardless' ) );
```
Not sure if this is needed, as subscriptions will get cancelled with payment cancelled webhook call, and order is already cancelled on line no 2316. am I missing anything here? Thanks

lmcrean replied:
Correct, WCS should handle subscription cancellation when the order status changes. Removed the entire method and the call to it. Much cleaner now.

iamdharmesh reviewed (changes requested):
Thanks a lot for the changes @lmcrean and sorry for the delayed review. I was on vacation, just came back this week. Changes now looks good. However, added some feedback to check. Could you please check it once. Thanks.

lmcrean commented:
@iamdharmesh thanks as always for the codebase insights. I have addressed all points made; see resolved comments. 

Ready for next review.

vikrampm1 commented:
@iamdharmesh this is back for your re-review when you get a moment.

iamdharmesh reviewed (approved):
Thanks a lot for the PR @lmcrean. I have made some minor changes and this is good to go now.

qasumitbagthariya commented:
Regression / Smoke Test Report  ✅ 
-------

Tested with Archive File created via "php woorelease.phar build repo_URL" (Composer version 2.8.4, npm version 10.8.2, node version 20.18.1)

Status- Working expected with Plugin Archive/Zip file same as fix specific branch.

https://github.com/user-attachments/assets/4bea3c5a-f72b-4fc7-b820-a0fe12d994ba

**Testing Environment** 

<Details>

- WordPress: 6.9
- Theme: Storefront 4.6.2
- Theme: Twenty Twenty-Five 1.4
- WooCommerce - 10.4.0
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS 15.2 
- Branch: smoke-testing

</Details>

Next Step- Ready to Merge 🚀
